### DISCUSSION TITLE
Integrating Tailwind CSS Components with NiceGUI

### DISCUSSION THREAD

**Discussion by Xtreemrus** [zauberzeug/nicegui#2337] [Q&A] [answered]
### Question

Hello everyone,

I've spent several hours trying to figure out how to work with Tailwind CSS components in NiceGUI, but I haven't come up with a solution yet.

Here's an example from the Tailwind CSS playground:
https://play.tailwindcss.com/foLEE0kfJn?size=968x720

```CSS
@import url("https://fonts.googleapis.com/css2?family=Roboto&display=swap");

@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
  .wrapper {
    @apply font-sans bg-purple-200 p-12 text-center;
  }
  .menu {
    @apply bg-white p-2 rounded-full;
  }
  a {
    @apply bg-blue-50 text-gray-800 py-2 px-6 rounded-full;
  }
  a.active {
    @apply bg-blue-500 text-white;
  }
}
```

I'm hesitant to use "Tailwind().apply" directly in Python or to insert the entire string into the class attribute because it defeats the purpose of HTML classes. Debugging the frontend in the browser would be more difficult since, instead of class names like 'wrapper', we would see a list of individual styles like "font-sans bg-purple-200 p-12 text-center".

Has anyone successfully integrated Tailwind CSS components with NiceGUI?

Thank you for your help!

**Comment by stepheweffie** [comment]
`    ui.button('Submit', classes='bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded')
`
The 'classes' parameter should allow Tailwind classes, e.g. w-full, w-1/2, h-screen etc.

**Reply by Xtreemrus** [reply]
Thank you for your response. However, I believe there might be a slight misunderstanding regarding my question. The solution you've provided involves using Tailwind's utility classes directly within the 'classes' parameter. While this certainly works for applying individual style classes, my aim is to integrate Tailwind's component layer, which allows for a more semantic and maintainable approach by defining custom components within the CSS.

For example, in the Tailwind CSS I've shared, I have a `.wrapper` component that encapsulates several Tailwind utility classes. I'm looking for a way to use these defined components directly in NiceGUI, rather than listing out all the utility classes in the 'classes' parameter.

**Reply by stepheweffie** [reply]
[Read the docs ](https://nicegui.io/documentation/section_styling_appearance#tailwind_css)
You can also apply [Tailwind CSS](https://tailwindcss.com/) utility classes with the classes method.

**Reply by falkoschindler** [reply]
@stepheweffie As far as I understand this doesn't answer @Xtreemrus' question, which is about combining Tailwind _classes_ into some kind of meta classes that can be used in NiceGUI.

**Comment by falkoschindler** [comment]
I'm not sure if this is possible with NiceGUI, but I'm no Tailwind expert. I would assume that we need some kind of compile step to translate the example code to plain CSS which can be served as part of the HTML page. If this can be done in Python, we might be able to build an API for it.

**Comment by Xtreemrus** [comment]
I mean this part of tailwind https://tailwindcss.com/docs/functions-and-directives#layer.
It's more like LESS.

You can see how it works in the sandbox example, the style is set in CSS as a class, but you can use tailwind syntax.
https://play.tailwindcss.com/foLEE0kfJn?size=968x720

You can see the difference in the generated HTML code, you can also see it in the inspector.

**Comment by falkoschindler** [comment, marked as answer]
I just noticed that this simply works:

```py
ui.add_head_html('''
    <style type="text/tailwindcss">
        @layer components {
            .blue-box {
                @apply bg-blue-500 p-12 text-center shadow-lg rounded-lg text-white;
            }
        }
    </style>
''')

ui.label('Hello').classes('blue-box')
ui.label('world').classes('blue-box')
```

<img width="305" alt="Screenshot 2024-01-12 at 07 47 43" src="https://github.com/zauberzeug/nicegui/assets/5767091/ced647ee-a633-46de-a360-462af1472510">

**Reply by Xtreemrus** [reply]
That's fantastic, thank you for sharing this solution! It's great to see that custom Tailwind CSS components can be integrated directly into NiceGUI in such a straightforward manner.

The use of classes is indeed considered a best practice for maintaining clean and manageable code, and it's not commonly known that one can continue to utilize them even when working with a utility-first framework like Tailwind CSS.
Your excellent example could serve as a valuable addition to the documentation, helping others who might be looking to implement a similar setup :)
